File: src-tauri/src/parser/parsers/match_metadata.rs
// Match-metadata pure-parse module.
//
// Parity-only counterpart to the legacy `import::insert_match_metadata`
// (src-tauri/src/parser/import.rs:1134–1282). Returns a serializable struct
// matching the TypeScript `MatchMetadata` envelope shape used by the cloud
// rewrite (`src/lib/parser/parsers/match-metadata.ts`).
//
// The desktop `insert_match_metadata` path is intentionally NOT refactored
// to call this — desktop is a working production deployment and we keep
// risk to that flow at zero. The three private helpers below
// (parse_version_string, parse_save_date, parse_game_content) are
// duplicated from import.rs; resync if any of those change upstream.

use crate::parser::game_data::PlayerData;
use crate::parser::xml_loader::{XmlDocument, XmlNodeExt};
use crate::parser::{ParseError, Result};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MatchMetadataStruct {
    pub xml_game_id: String,
    pub total_turns: i32,
    pub game_name: Option<String>,
    pub save_date: Option<String>,
    pub game_version: Option<String>,
    pub map_width: Option<i32>,
    pub map_height: Option<i32>,
    pub map_size: Option<String>,
    pub map_class: Option<String>,
    pub game_mode: Option<String>,
    pub difficulty: Option<String>,
    pub opponent_level: Option<String>,
    pub victory_conditions: Option<String>,
    pub enabled_mods: Option<String>,
    pub enabled_dlc: Option<String>,
    pub winner: Option<WinnerInfoStruct>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WinnerInfoStruct {
    pub winner_player_xml_id: i32,
    pub winner_team_id: Option<i32>,
    pub victory_type: String,
}

/// Parse match metadata + resolved winner. The `players` slice must already
/// be available — winner resolution can need to map `<TeamVictories>` team
/// id → owning player via Player @Team attributes.
pub fn parse_match_metadata_struct(
    doc: &XmlDocument,
    players: &[PlayerData],
) -> Result<MatchMetadataStruct> {
    let root = doc.root_element();
    let xml_game_id = root.req_attr("GameId")?.to_string();

    let game_node = root
        .children()
        .find(|n| n.has_tag_name("Game"))
        .ok_or_else(|| ParseError::MissingElement("Game".to_string()))?;
    let total_turns: i32 = game_node
        .opt_child_text("Turn")
        .and_then(|t| t.parse().ok())
        .ok_or_else(|| ParseError::MissingElement("Game.Turn".to_string()))?;

    let (game_version, enabled_mods) = root
        .opt_attr("Version")
        .map(parse_version_string)
        .unwrap_or((None, None));
    let enabled_dlc = parse_game_content(&root);
    let save_date = root.opt_attr("SaveDate").and_then(parse_save_date);

    let map_width: Option<i32> = root.opt_attr("MapWidth").and_then(|s| s.parse().ok());
    // MapHeight isn't encoded; the legacy code assumes square map.
    let map_height = map_width;

    let game_name = root
        .opt_attr("GameName")
        .or_else(|| root.opt_child_text("GameName"))
        .map(|s| s.to_string());

    let victory_conditions = root
        .children()
        .find(|n| n.has_tag_name("VictoryEnabled"))
        .map(|ve| {
            ve.children()
                .filter(|c| c.is_element())
                .map(|c| c.tag_name().name().to_string())
                .collect::<Vec<_>>()
                .join("+")
        })
        .filter(|s| !s.is_empty());

    let team_assignments: Vec<i32> = root
        .children()
        .find(|n| n.has_tag_name("Team"))
        .map(|team_elem| {
            team_elem
                .children()
                .filter(|n| n.has_tag_name("PlayerTeam"))
                .filter_map(|n| n.text().and_then(|t| t.parse::<i32>().ok()))
                .collect()
        })
        .unwrap_or_default();

    // Winner detection — same two-source strategy as the desktop:
    //   1. <Game><TeamVictories><Team Victory="...">team_id</Team>
    //   2. <Victory winner="player_xml_id" type="...">
    let winner = detect_winner(&root, &game_node, &team_assignments, players);

    Ok(MatchMetadataStruct {
        xml_game_id,
        total_turns,
        game_name,
        save_date,
        game_version,
        map_width,
        map_height,
        map_size: root.opt_attr("MapSize").map(|s| s.to_string()),
        map_class: root.opt_attr("MapClass").map(|s| s.to_string()),
        game_mode: root.opt_attr("GameMode").map(|s| s.to_string()),
        // Difficulty isn't a Root attribute in this XML shape; mirrors the
        // TS parser which reads `@_Difficulty` and yields null. Kept for
        // shape symmetry with the cloud blob's MatchMetadata.
        difficulty: root.opt_attr("Difficulty").map(|s| s.to_string()),
        opponent_level: root.opt_attr("OpponentLevel").map(|s| s.to_string()),
        victory_conditions,
        enabled_mods,
        enabled_dlc,
        winner,
    })
}

fn detect_winner(
    root: &roxmltree::Node,
    game_node: &roxmltree::Node,
    team_assignments: &[i32],
    players: &[PlayerData],
) -> Option<WinnerInfoStruct> {
    // Source 1: <Game><TeamVictories><Team Victory="...">winning_team_id</Team>
    if let Some(tv) = game_node.children().find(|n| n.has_tag_name("TeamVictories")) {
        if let Some(team) = tv.children().find(|n| n.has_tag_name("Team")) {
            let team_id = team.text().and_then(|t| t.parse::<i32>().ok());
            let victory_type = team.opt_attr("Victory").map(|s| s.to_string());
            if let (Some(team_id), Some(victory_type)) = (team_id, victory_type) {
                if let Some(player_xml_id) =
                    resolve_player_for_team(team_id, team_assignments, players)
                {
                    return Some(WinnerInfoStruct {
                        winner_player_xml_id: player_xml_id,
                        winner_team_id: Some(team_id),
                        victory_type,
                    });
                }
            }
        }
    }

    // Source 2: <Victory winner="player_xml_id" type="...">
    if let Some(v) = root.children().find(|n| n.has_tag_name("Victory")) {
        let winner_id = v.opt_attr("winner").and_then(|s| s.parse::<i32>().ok());
        let victory_type = v.opt_attr("type").map(|s| s.to_string());
        if let (Some(winner_id), Some(victory_type)) = (winner_id, victory_type) {
            return Some(WinnerInfoStruct {
                winner_player_xml_id: winner_id,
                winner_team_id: None,
                victory_type,
            });
        }
    }

    None
}

fn resolve_player_for_team(
    team_id: i32,
    team_assignments: &[i32],
    players: &[PlayerData],
) -> Option<i32> {
    // Strategy 1: index-by-position in <Team><PlayerTeam>… (Player XML id
    // == position).
    if let Some(idx) = team_assignments.iter().position(|&t| t == team_id) {
        return Some(idx as i32);
    }

    // Strategy 2: Player @Team attribute equals team_id (as string).
    let team_id_str = team_id.to_string();
    if let Some(p) = players
        .iter()
        .find(|p| p.team_id.as_ref() == Some(&team_id_str))
    {
        return Some(p.xml_id);
    }

    // Strategy 3: SP fallback — single human player. Only safe when exactly
    // one human is present; otherwise we'd guess.
    let humans: Vec<&PlayerData> = players.iter().filter(|p| p.is_human).collect();
    if humans.len() == 1 {
        return Some(humans[0].xml_id);
    }

    None
}

// ---------- Private helpers (duplicated from import.rs) ----------

/// Format: `Version: 1.0.70671+MOD_NAME=hash+MOD_NAME2=hash`. Returns
/// `(version_number, joined_mod_names)`. Mirrors `parse_version_string` at
/// import.rs:889–913.
fn parse_version_string(version: &str) -> (Option<String>, Option<String>) {
    let parts: Vec<&str> = version.split('+').collect();
    if parts.is_empty() {
        return (None, None);
    }

    let version_num = parts[0]
        .strip_prefix("Version: ")
        .map(|v| v.to_string());

    let mods = if parts.len() > 1 {
        let mod_parts: Vec<&str> = parts[1..]
            .iter()
            .map(|s| s.split('=').next().unwrap_or(s))
            .collect();
        Some(mod_parts.join("+"))
    } else {
        None
    };

    (version_num, mods)
}

/// Walk `<GameContent>` for `DLC_*` children. Mirrors `parse_game_content`
/// at import.rs:918–933.
fn parse_game_content(root: &roxmltree::Node) -> Option<String> {
    let game_content = root.children().find(|n| n.has_tag_name("GameContent"))?;
    let dlcs: Vec<&str> = game_content
        .children()
        .filter(|child| child.is_element())
        .map(|el| el.tag_name().name())
        .filter(|name| name.starts_with("DLC_"))
        .collect();
    if dlcs.is_empty() {
        None
    } else {
        Some(dlcs.join("+"))
    }
}

/// "31 January 2024" → ISO `YYYY-MM-DD`. Mirrors `parse_save_date` at
/// import.rs:937–944.
fn parse_save_date(date_str: &str) -> Option<String> {
    use chrono::NaiveDate;
    NaiveDate::parse_from_str(date_str, "%d %B %Y")
        .ok()
        .map(|d| d.format("%Y-%m-%d").to_string())
}

File: src/lib/parser/parsers/match-metadata.ts
// Match metadata parser. Pure-parse port of
// `insert_match_metadata` (src-tauri/src/parser/import.rs:1134–1282) +
// `update_winner` (same file lines 956–1008) — the DB-write paths are
// dropped; this module returns a typed `MatchMetadata` struct.
//
// The parser depends on already-parsed `Player[]` from `players.ts` so it
// can resolve a winning-team-id back to a player XML id without a second
// pass over the XML.

import { ParseError } from "../extract-zip.js";
import {
	asArray,
	getElementChildren,
	isElement,
	optAttrStr,
	optInt,
	optStr,
	requireStr,
} from "../parse-xml.js";
import type { MatchMetadata, RawWinnerInfo, WinnerInfo } from "../types.js";
import type { Player } from "./players.js";

export function parseMatchMetadata(
	root: Record<string, unknown>,
	players: Player[],
): MatchMetadata {
	const xmlGameId = requireStr(root["@_GameId"], "Root.GameId");

	const gameNode = root.Game;
	if (!isElement(gameNode)) {
		throw new ParseError("Game", "MISSING_FIELD");
	}
	const totalTurns = optInt(gameNode.Turn);
	if (totalTurns === null) {
		throw new ParseError("Game.Turn", "MISSING_FIELD");
	}

	const versionAttr = optAttrStr(root["@_Version"]);
	const [gameVersion, enabledMods] = versionAttr
		? parseVersionString(versionAttr)
		: [null, null];

	const enabledDlc = parseGameContent(root);
	const saveDateAttr = optAttrStr(root["@_SaveDate"]);
	const saveDate = saveDateAttr ? parseSaveDate(saveDateAttr) : null;

	const mapWidth = optInt(root["@_MapWidth"]);
	// MapHeight is not encoded in the XML; the Rust parser assumes a square
	// map and reuses MapWidth. Keep the same convention.
	const mapHeight = mapWidth;

	const gameName = optAttrStr(root["@_GameName"]) ?? optStr(root.GameName);

	const victoryEnabled = parseVictoryEnabledList(root);
	const victoryConditions =
		victoryEnabled.length > 0 ? victoryEnabled.join("+") : null;
	const teamAssignments = parseTeamAssignments(root);

	const gameOver = "GameOver" in gameNode;

	const rawWinner = detectRawWinner(root, gameNode, victoryEnabled);
	const winner = rawWinner
		? resolveWinner(rawWinner, players, teamAssignments)
		: null;

	return {
		xml_game_id: xmlGameId,
		total_turns: totalTurns,
		game_name: gameName,
		save_date: saveDate,
		game_version: gameVersion,
		map_width: mapWidth,
		map_height: mapHeight,
		map_size: optAttrStr(root["@_MapSize"]),
		map_class: optAttrStr(root["@_MapClass"]),
		game_mode: optAttrStr(root["@_GameMode"]),
		// Difficulty is per-player in the save (`<Difficulty><PlayerDifficulty>`);
		// the match-level "Difficulty" stamp is the save owner's tier, set by
		// parsePlayers via the <?ActivePlayer N?> PI or the single-human fallback.
		// Returns null when no save owner could be identified (observer upload
		// of multi-human MP without the PI).
		difficulty: players.find((p) => p.isSaveOwner)?.difficulty ?? null,
		opponent_level: optAttrStr(root["@_OpponentLevel"]),
		victory_conditions: victoryConditions,
		enabled_mods: enabledMods,
		enabled_dlc: enabledDlc,
		game_over: gameOver,
		winner,
	};
}

// ---------- Version + mods ----------

/**
 * Parse the `Version` attribute. Format:
 *   `Version: 1.0.70671+MOD_NAME=hash+MOD_NAME2=hash`
 * Returns `[version_number, joined_mod_names]`. Mirrors
 * `parse_version_string` in import.rs:889–913.
 */
function parseVersionString(version: string): [string | null, string | null] {
	const parts = version.split("+");
	if (parts.length === 0) return [null, null];

	const versionNum = parts[0].startsWith("Version: ")
		? parts[0].slice("Version: ".length)
		: null;

	if (parts.length === 1) return [versionNum, null];

	// Each mod entry is "MOD_NAME=hash" — drop the hash.
	const modNames = parts.slice(1).map((s) => s.split("=")[0]);
	return [versionNum, modNames.join("+")];
}

// ---------- DLC content ----------

/**
 * Walk `<GameContent>` children for `DLC_*` element names. Mirrors
 * `parse_game_content` in import.rs:918–933.
 */
function parseGameContent(root: Record<string, unknown>): string | null {
	const gameContent = root.GameContent;
	if (!isElement(gameContent)) return null;

	const dlcs: string[] = [];
	for (const [name] of getElementChildren(gameContent)) {
		if (name.startsWith("DLC_")) dlcs.push(name);
	}
	return dlcs.length > 0 ? dlcs.join("+") : null;
}

// ---------- Save date ----------

const SAVE_DATE_RE = /^(\d{1,2}) ([A-Za-z]+) (\d{4})$/;
const MONTH_INDEX: Record<string, number> = {
	January: 1,
	February: 2,
	March: 3,
	April: 4,
	May: 5,
	June: 6,
	July: 7,
	August: 8,
	September: 9,
	October: 10,
	November: 11,
	December: 12,
};

/**
 * Parse "31 January 2024"-style strings to ISO `YYYY-MM-DD`. Returns null
 * for unparseable input — same lenient behavior as Rust's `chrono::NaiveDate
 * ::parse_from_str(...).ok()`.
 */
function parseSaveDate(dateStr: string): string | null {
	const m = SAVE_DATE_RE.exec(dateStr.trim());
	if (!m) return null;
	const day = parseInt(m[1], 10);
	const month = MONTH_INDEX[m[2]];
	const year = parseInt(m[3], 10);
	if (Number.isNaN(day) || month === undefined || Number.isNaN(year)) {
		return null;
	}
	const dayPad = day.toString().padStart(2, "0");
	const monthPad = month.toString().padStart(2, "0");
	return `${year}-${monthPad}-${dayPad}`;
}

// ---------- Victory conditions ----------

/**
 * Ordered list of enabled victory types from `<VictoryEnabled>`. Two consumers:
 *   1. `victory_conditions` field — joined with `+` for compact storage.
 *   2. Mapping a `<WinnerVictory>` integer to a `VICTORY_*` string for the
 *      legacy winner format (older save versions). The integer is a runtime
 *      info-list index; this list is the per-save snapshot of that order.
 */
function parseVictoryEnabledList(root: Record<string, unknown>): string[] {
	const ve = root.VictoryEnabled;
	if (!isElement(ve)) return [];

	const names: string[] = [];
	for (const [name] of getElementChildren(ve)) names.push(name);
	return names;
}

// ---------- Team assignments ----------

/**
 * `<Team><PlayerTeam>0</PlayerTeam><PlayerTeam>1</PlayerTeam></Team>` —
 * the index of each `PlayerTeam` corresponds to the player XML id and its
 * text is that player's team id. Returns the array indexed by player XML id.
 */
function parseTeamAssignments(root: Record<string, unknown>): number[] {
	const teamElem = root.Team;
	if (!isElement(teamElem)) return [];

	const out: number[] = [];
	for (const child of asArray(teamElem.PlayerTeam) as unknown[]) {
		// Children may be primitive strings (text-only) or element objects.
		// fast-xml-parser only wraps when ALWAYS_ARRAY_TAGS includes the name;
		// PlayerTeam is in the list, so we always get an array.
		const text = typeof child === "string" ? child : null;
		if (text === null) continue;
		const teamId = parseInt(text, 10);
		if (Number.isNaN(teamId)) continue;
		out.push(teamId);
	}
	return out;
}

// ---------- Winner detection ----------

/**
 * Pull a raw winner from XML. Three sources, in priority order:
 *   1. `<Game><TeamVictories><Team Victory="...">winning_team_id</Team></TeamVictories>`
 *      — newer save format (≥ ~1.0.79513).
 *   2. `<Game><WinnerTeam>N</WinnerTeam><WinnerVictory>N</WinnerVictory>`
 *      — legacy save format (≤ ~1.0.70671). `WinnerVictory` is an integer
 *      index into the per-save `<VictoryEnabled>` list, passed in here.
 *   3. `<Victory winner="player_xml_id" type="...">` at the root — older
 *      single-player fallback shape (rarely seen in modern saves).
 * Returns the team-id form for (1)/(2) and the player-id form for (3). The
 * resolver below maps either back to a player.
 */
function detectRawWinner(
	root: Record<string, unknown>,
	gameNode: Record<string, unknown>,
	victoryEnabled: string[],
): RawWinnerInfo | { player_xml_id: number; victory_type: string } | null {
	const teamVictories = gameNode.TeamVictories;
	if (isElement(teamVictories)) {
		// `<TeamVictories>` may contain multiple `<Team>` children when more
		// than one team has met victory criteria simultaneously. Mirrors
		// Rust's `tv.children().find(|n| n.has_tag_name("Team"))` — first
		// match wins. Note: "Team" is not in ALWAYS_ARRAY_TAGS, so a single
		// child arrives as an object and multiple as an array; asArray
		// normalizes both shapes.
		const teamEntries = asArray(teamVictories.Team) as unknown[];
		for (const teamEntry of teamEntries) {
			if (!isElement(teamEntry)) continue;
			const teamIdRaw = optInt(teamEntry["#text"]);
			const victoryType = optAttrStr(teamEntry["@_Victory"]);
			if (teamIdRaw !== null && victoryType !== null) {
				return { team_id: teamIdRaw, victory_type: victoryType };
			}
		}
	}

	// Legacy format: <WinnerTeam>int</WinnerTeam> + <WinnerVictory>int</WinnerVictory>.
	// Both elements are present on completed games in older versions and absent
	// in newer versions — checking both ensures we don't misread a default-zero
	// integer when only one is set.
	const winnerTeamId = optInt(gameNode.WinnerTeam);
	const winnerVictoryIdx = optInt(gameNode.WinnerVictory);
	if (winnerTeamId !== null && winnerVictoryIdx !== null) {
		const victoryType = victoryEnabled[winnerVictoryIdx];
		if (victoryType !== undefined) {
			return { team_id: winnerTeamId, victory_type: victoryType };
		}
	}

	const victoryRoot = root.Victory;
	if (isElement(victoryRoot)) {
		const winnerId = optInt(victoryRoot["@_winner"]);
		const victoryType = optAttrStr(victoryRoot["@_type"]);
		if (winnerId !== null && victoryType !== null) {
			return { player_xml_id: winnerId, victory_type: victoryType };
		}
	}

	return null;
}

/**
 * Resolve a raw winner record into a `WinnerInfo` with a concrete player
 * XML id. Three strategies, in order (combines insert_match_metadata's
 * team_assignments lookup with update_winner's two fallbacks):
 *   1. team_assignments[player_xml_id] === team_id (Team/PlayerTeam mapping)
 *   2. players[i].teamId === team_id_str (Player @Team attribute)
 *   3. the single human player (single-player save fallback)
 *
 * If the raw winner already carries a player_xml_id (Victory-element form),
 * returns it as-is.
 */
function resolveWinner(
	raw: RawWinnerInfo | { player_xml_id: number; victory_type: string },
	players: Player[],
	teamAssignments: number[],
): WinnerInfo | null {
	if ("player_xml_id" in raw) {
		return {
			winner_player_xml_id: raw.player_xml_id,
			winner_team_id: null,
			victory_type: raw.victory_type,
		};
	}

	const teamId = raw.team_id;
	if (teamId === null) return null;

	// Strategy 1: Team/PlayerTeam mapping.
	const idxByTeam = teamAssignments.indexOf(teamId);
	if (idxByTeam !== -1) {
		return {
			winner_player_xml_id: idxByTeam,
			winner_team_id: teamId,
			victory_type: raw.victory_type,
		};
	}

	// Strategy 2: Player @Team attribute.
	const teamIdStr = teamId.toString();
	const byAttr = players.find((p) => p.teamId === teamIdStr);
	if (byAttr) {
		return {
			winner_player_xml_id: byAttr.xmlId,
			winner_team_id: teamId,
			victory_type: raw.victory_type,
		};
	}

	// Strategy 3: SP fallback — the single human player. Only safe when
	// exactly one human exists; otherwise we'd guess. Mirrors update_winner's
	// "find the human player" path at import.rs:976.
	const humans = players.filter((p) => p.isHuman);
	if (humans.length === 1) {
		return {
			winner_player_xml_id: humans[0].xmlId,
			winner_team_id: teamId,
			victory_type: raw.victory_type,
		};
	}

	// No resolution — surface a parse error rather than silently dropping a
	// winner detection. The caller can downgrade this to `winner: null` if
	// the policy is "incomplete game" rather than "broken save".
	throw new ParseError(
		`Could not resolve winning team ${teamId} to a player`,
		"INVALID_FORMAT",
	);
}
